### ISSUE TITLE
Running Wizard as a non-modal/non-popup is broken

### ISSUE THREAD

**Issue by GaikwadPratik** [migueldeicaza/gui.cs#1849]
Hi,

Following structure of `UICatalog`, I am trying to run `Wizard` scenario. The difference is that **instead of** calling `Application.Run` on `Wizard` as done [here](https://github.com/migueldeicaza/gui.cs/blob/main/UICatalog/Scenarios/Wizards.cs#L286), I am simply adding it inside the [MainWindow](https://github.com/migueldeicaza/gui.cs/blob/main/UICatalog/Scenario.cs#L58) which is included in the `Top`.  In this case, when the wizard shows the last step added as first step and then behaves completely in incorrect way. But if I call `Application.Run(wizard)` it works properly though it doesn't show any other controls, like menu, that I have added in `Top`.

I have updated my sample app [TerrminalWizard](https://github.com/GaikwadPratik/TerrminalWizard)

What am I missing?

https://user-images.githubusercontent.com/5595263/177230377-9a6a4530-2a7d-4d7d-b6d5-52c725435770.mov

**Comment by tig** [comment]
I’ll take a look tomorrow. I suspect it’s a bug. I bet Dialog fails the same way.

-cek
Sent via mobile. Expect brevity, typos, and dangerous driving.
________________________________
From: Gaikwad Pratik ***@***.***>
Sent: Monday, July 4, 2022 7:01:42 PM
To: migueldeicaza/gui.cs ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [migueldeicaza/gui.cs] Running Wizard inside window scrambles steps (Issue #1849)


Hi,

Following structure of UICatalog, I am trying to run Wizard scenario. The difference is that instead of calling Application.Run on Wizard as done here<https://github.com/migueldeicaza/gui.cs/blob/main/UICatalog/Scenarios/Wizards.cs#L286>, I am simply adding it inside the MainWindow<https://github.com/migueldeicaza/gui.cs/blob/main/UICatalog/Scenario.cs#L58> which is included in the Top. In this case, when the wizard shows the last step added as first step and then behaves completely in incorrect way. But if I call Application.Run(wizard) it works properly though it doesn't show any other controls, like menu, that I have added in Top.

I have updated my sample app TerrminalWizard<https://github.com/GaikwadPratik/TerrminalWizard>

What am I missing?

**Comment by tig** [comment]
Just because I was bored, I coded it up to work as a non-modal. I personally don't think it makes much sense to do it like this, but if you want, I'll submit a PR to fix this. Let me know.

![a0yCxjA 1](https://user-images.githubusercontent.com/585482/177453776-365075c5-9aa6-4b43-8bd3-27bde65f3661.gif)

To make it work like this, you'll adjust your code to start the wizard thusly:

```csharp
					if (modalCheckBox.Checked) {
						Application.Run (wizard);
					} else {
						// Disable the Show button so this only happens once
						showWizardButton.Visible = false;
						// To use Wizard as a View, you must set Modal = false
						wizard.Modal = false;
						// Turn off borders since it's not a popup
						wizard.Border.Effect3D = false;
						wizard.Border.BorderStyle = BorderStyle.None;
						wizard.ColorScheme = Win.ColorScheme;
						Win.Add (wizard);
					}
```

**Comment by GaikwadPratik** [comment]
@tig,

I am going to use this wizard during set up of linux server installation. Similar to what you see in any Linux distro or Windows installation. It will be asking users a series of inputs. In case, user decided to cancel the installation or restart server in between, with current setup, they will have to first use `ctrl + c` to stop `Wizard` and then use `menu` to do exit main application which would be slightly confusing for non developers. They will have to remember and perform the extra step of `ctrl + c` even if it is mentioned in the documentation. 

But with your PR changes, it would simply to use `menu` to exit main application. 

P.S. I am sure, there will be many more such use cases, where in it would be useful to NOT have `wizard` and `dialog` as modal.

**Comment by tig** [comment]
> @tig ,
> 
> Any thoughts?

I've been working on it. Along the way I've discovered a bug that I'm trying to track down. 

Why do you need a menu?

Why not have 'Wizard' be the only 'TopLevel' of the app. So that if the user presses 'ctrl-c' (or 'ctrl-q' which is the default `QuitKey`) the Wizard (and your app) exits?

**Comment by GaikwadPratik** [comment]
> Why not have 'Wizard' be the only 'TopLevel' of the app. So that if the user presses 'ctrl-c' (or 'ctrl-q' which is the default QuitKey) the Wizard (and your app) exits?

I can probably make that work. But I feel like having a familiar option might be easier

I actually wanted them to present with 3 options in the menu:

1. Restart the wizard app(restart configuration)
2. Restart the server(go through OS installation process again)
3. Shutdown the server(quit the installer and do not use the product/OS at all)

If they press `ctrl-q`, then I am not sure what action the user will want to perform from 3 options above.

**Comment by tig** [comment]
![307Jfsm 1](https://user-images.githubusercontent.com/585482/179437859-26c1da31-d1bc-4f35-a4c2-f8ce994ab689.gif)

How's this look?

Note, I'm still playing with some layout and color things. Prepping a PR now.

**Comment by GaikwadPratik** [comment]
This would actually work. Thank you very much for working on this and seeing this through.

**Comment by GaikwadPratik** [comment]
@tig,

Can you please point out to sample code that generated above image? Is it achieved by just setting `wizard.Model = false` or anything else as well?

**Comment by BDisp** [comment]
In `Wizard` scenario on the line where is:
```cs
Application.Run (wizard);
```
replace by:


> ```cs
> 					if (modalCheckBox.Checked) {
> 						Application.Run (wizard);
> 					} else {
> 						// Disable the Show button so this only happens once
> 						showWizardButton.Visible = false;
> 						// To use Wizard as a View, you must set Modal = false
> 						wizard.Modal = false;
> 						// Turn off borders since it's not a popup
> 						wizard.Border.Effect3D = false;
> 						wizard.Border.BorderStyle = BorderStyle.None;
> 						wizard.ColorScheme = Win.ColorScheme;
> 						Win.Add (wizard);
> 					}
> ```

**Comment by GaikwadPratik** [comment]
Makes sense. I asked because it is commented in https://github.com/migueldeicaza/gui.cs/pull/1853/commits/9445f62b58fc6ad466560d68e15bac81a89d1bf3#diff-c7a1268c8538e116c01d122984a9c73de870fa0545a3912a3532c22d7c3b6eb7R21

**Comment by GaikwadPratik** [comment]
Thank you for clarification. Just waiting for PR to be merged and having more fun developing the app.

**Comment by GaikwadPratik** [comment]
@tig ,

This is still not working for me. 

1. It shows lastly added step as first with `back` and `finished` buttons. Upon clicking `finished` button, it shows first step and behaves correctly. Irrespective of whether I am setting `Modal` property.
2. If I do NOT set `Modal = false` then it shows background as Grey, but if I set it, then it's not using the color scheme that I am setting. Is this expected? I am setting `Colors.Dialog` in the constructor of wizard.
3. What is the sequence of events on `finished` button click? `StepChanging` will be fired or not? If not, how do I read the values of input controls in the last step? Should I be using `MovingBack` and `MovingNext` events instead of `StepChanging` event?

P.S. I have updated my sample app [TerrminalWizard](https://github.com/GaikwadPratik/TerrminalWizard) and all the above points are observed upon pulling latest code changes till today.

**Comment by tznind** [comment]
> I am setting `Colors.Dialog` in the constructor of wizard.

Typically you want to avoid manually setting `ColorScheme` on an object unless you are actually trying to achieve a color change.  The default behaviour of a view is to inherit from it's parent.

So if you have a `Window` (defaults to `Color.Base`) and add a non modal View (button, label, wizard etc).  Then the view will automatically take its parent color scheme.

If you have a `Dialog` (defaults to `Color.Dialog` - greyish background) and add a non modal view (button, label, wizard etc) then the view will take the parent scheme again and have a grey colored background (same as parent).

**Comment by tig** [comment]
Looking at `TerminalWizard`, i'm confused why you are re-using all the scaffolding from `Scenario.cs`. It is adding a ton of complexity to your app for no reason I can see. I believe it is masking what is really going on.

**Comment by tig** [comment]
I found your issue. When using Wizard as a non-modal, you must explicitly set the first step. This is because when running as a modal, the `Loaded` event is fired, and I can set the first step there, but when non-modals don't get a `Loaded` event.

```csharp
			// When run as a modal, Wizard gets a Loading event where it sets the
			// Current Step. But when running non-modal it must be done manually.
			wizard.CurrentStep = wizard.GetNextStep ();
```

I should have noted this in the API documentation for Wizard; I will submit a PR to do this asap.

I've submitted a PR to your project with this fix and with some re-factoring to simplify your code.  I also couldn't resist cleaning some other stuff up...hope you don't mind!

**Comment by BDisp** [comment]
> I found your issue. When using Wizard as a non-modal, you must explicitly set the first step. This is because when running as a modal, the `Loaded` event is fired, and I can set the first step there, but when non-modals don't get a `Loaded` event.

Maybe you can use the `Initialized` event instead but I'm not sure.

**Comment by tig** [comment]
> 3. What is the sequence of events on finished button click? StepChanging will be fired or not? If not, how do I read the values of input controls in the last step? Should I be using MovingBack and MovingNext events instead of StepChanging event?

This is a good question.  

`StepChanging`/`StepChanged`/`MovingBack`/`MovingNext` are only fired when a step transition is happening. When the user clicks the Next/Finished button on the last step they do not get fired. However the `Finished` event is fired and you can read the values in that event handler. 

Some devs will want to capture/save/analyize form data at each step transition. They should typically do it on `StepChanging` for all steps except the last. For the last step, they should do it on `Finished`.

In some cases, devs will just want to capture everything at the very end. They can just handle `Finished`.

Because you want to be able to save state if the user cancels when in the middle of the wizard, you'll want to do the former.

**Comment by tig** [comment]
> > I found your issue. When using Wizard as a non-modal, you must explicitly set the first step. This is because when running as a modal, the `Loaded` event is fired, and I can set the first step there, but when non-modals don't get a `Loaded` event.
> 
> Maybe you can use the `Initialized` event instead but I'm not sure.

Nope. It doesn't get called either for non-Modals.

**Comment by GaikwadPratik** [comment]
> Looking at `TerminalWizard`, i'm confused why you are re-using all the scaffolding from `Scenario.cs`. It is adding a ton of complexity to your app for no reason I can see. I believe it is masking what is really going on.

I am also going to need to load couple of more views once the configuration is finished within the wizard. That is one more reason, I was hoping for non-modal wizard. More on that later, once I am through FBWizard. :D

**Comment by GaikwadPratik** [comment]
> I found your issue. When using Wizard as a non-modal, you must explicitly set the first step. This is because when running as a modal, the `Loaded` event is fired, and I can set the first step there, but when non-modals don't get a `Loaded` event.
> 
> ```cs
> 			// When run as a modal, Wizard gets a Loading event where it sets the
> 			// Current Step. But when running non-modal it must be done manually.
> 			wizard.CurrentStep = wizard.GetNextStep ();
> ```
> 
> I should have noted this in the API documentation for Wizard; I will submit a PR to do this asap.
> 
> I've submitted a PR to your project with this fix and with some re-factoring to simplify your code. I also couldn't resist cleaning some other stuff up...hope you don't mind!

Thank you so much. This helps a lot. Hopefully I should be able to finish the wizarding part of my app based on this during this weekend.

**Comment by tig** [comment]
> > Looking at `TerminalWizard`, i'm confused why you are re-using all the scaffolding from `Scenario.cs`. It is adding a ton of complexity to your app for no reason I can see. I believe it is masking what is really going on.
> 
> I am also going to need to load couple of more views once the configuration is finished within the wizard. That is one more reason, I was hoping for non-modal wizard. More on that later, once I am through FBWizard. :D

Fine, but I recommend you avoid all the complexity of `Scenario`. It was built the way it was to enable lots of developers to add Scenarios to UICatalog over time. You don't need that.

**Comment by GaikwadPratik** [comment]
> > > Looking at `TerminalWizard`, i'm confused why you are re-using all the scaffolding from `Scenario.cs`. It is adding a ton of complexity to your app for no reason I can see. I believe it is masking what is really going on.
> > 
> > 
> > I am also going to need to load couple of more views once the configuration is finished within the wizard. That is one more reason, I was hoping for non-modal wizard. More on that later, once I am through FBWizard. :D
> 
> Fine, but I recommend you avoid all the complexity of `Scenario`. It was built the way it was to enable lots of developers to add Scenarios to UICatalog over time. You don't need that.

So I should use `Top.Remove` followed by `Top.Add` instead?

**Comment by tig** [comment]
I'd just use `.Visible`

**Comment by GaikwadPratik** [comment]
> I'd just use `.Visible`

Okay. That makes sense. One last question for the day, In your PR to `TerrminalWizard`, you have inherited `FirstBootWizard` from `Wizard` because of #1883. Once #1883 is resolved, the inheritance becomes unnecessary correct?

**Comment by BDisp** [comment]
> Nope. It doesn't get called either for non-Modals. I'm going to submit an issue on this.

Yes is true, it only be invoked at Application.Begin method. For views added afterwards they won't benefice it. Another event that occur to me is the Added that is invoked whatever a view is added to the super-view.

**Comment by tig** [comment]
Nope. Unrelated. I just did that because you already had a class and it was convenient vs having an ‘Wizard’ member variable.

-cek
Sent via mobile. Expect brevity, typos, and dangerous driving.
________________________________

**Comment by tig** [comment]
@GaikwadPratik - In my fix for #1883 I have decided to make it so setting the first step explicitly is not needed IN YOUR CASE.

There is still a case where devs will need to do it, but it's an outlier. The docs will now read:

<img width="553" alt="image" src="https://user-images.githubusercontent.com/585482/180616815-01916d7d-6303-4989-bfab-d6add7947fec.png">

**Comment by GaikwadPratik** [comment]
@tig,

One more question on this, just for my understanding. Why can't Wizard control(I have not used Dialog but I am sure it would be same) be like other controls which can be used in any other views like `Frame` or `Window`? Why does it have to be either `TopLevel`(modal) or can be added only in `Top` view?

P.S. This question is to just get an understanding.
